### PR TITLE
buildBlockExtrinsics to continue when apply fails

### LINKED ISSUE #1225: babe buildBlockExtrinsics should continue if applying an extrinsic fails
## Expected Behavior

- if applying an extrinsic fails while building a block, the node should ignore that extrinsic and continue to build
- however it errors if one extrinsic fails (see `lib/babe/build.go` `buildBlockExtrinsics`)
- the flow should be: pop extrinsic from queue -> applyExtrinsic -> continue to next extrinsic (essentially ignore / only log failures)

- Testing blocked by extrinsics #1211 (i..e create an additional issue for testing)

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [ ] I am planning to submit a pull request to fix this issue myself

### PR DESCRIPTION
**Pull request by RyRy79261**
## Changes

- Removed error returns from buildBlockExtrinsics loop

## Tests

```
go test ./lib/babe/... -short
go test ./... -short
```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [ ] I have reviewed my own pull request before requesting a review
- [ ] All integration tests and required coverage checks are passing

## Issues

- closes #1225

### CODE REVIEW

File: lib/babe/build.go [hunk 1]
```diff
@@ -187,24 +187,24 @@ func (b *Service) buildBlockExtrinsics(slot Slot) ([]*transaction.ValidTransacti
...
+			// return nil, err
 		}
 
 		// if ret == 0x0001, there is a dispatch error; if ret == 0x01, there is an apply error
 		if ret[0] == 1 || bytes.Equal(ret[:2], []byte{0, 1}) {
-			errTxt, err := determineError(ret)
-			if err != nil {
-				return nil, err
-			}
+			_, err := determineError(ret)
```
Comment (noot) [inline lib/babe/build.go:198]: you need the errTxt

Comment (noot) [reply lib/babe/build.go:198]: it tells you why applying failed

File: lib/babe/build.go [hunk 2]
```diff
@@ -187,24 +187,24 @@ func (b *Service) buildBlockExtrinsics(slot Slot) ([]*transaction.ValidTransacti
...
 			// remove invalid extrinsic from queue
 			b.transactionState.Pop()
+			if err == nil {
+				// re-add previously popped extrinsics back to queue
+				b.addToQueue(included)
+			}
```
Comment (noot) [inline lib/babe/build.go:204]: you don't need to re add, you can move the pop to the start of the for loop

File: lib/babe/build.go [hunk 3]
```diff
@@ -180,42 +177,38 @@ func (b *Service) buildBlockBabeHeader(slot Slot) (*types.BabeHeader, error) {
...
 			// remove invalid extrinsic from queue
-			b.transactionState.Pop()
-
-			// re-add previously popped extrinsics back to queue
-			b.addToQueue(included)
-
-			return nil, errors.New("error applying extrinsic: " + errTxt)
+			if err == nil {
+				b.addToQueue(included)
+			}
```
Comment (noot) [inline lib/babe/build.go:199]: please remove this, if the extrinsic is invalid we shouldn't re-add it back to the queue

File: lib/babe/build.go [hunk 4]
```diff
@@ -180,42 +177,38 @@ func (b *Service) buildBlockBabeHeader(slot Slot) (*types.BabeHeader, error) {
...
 
 	for !hasSlotEnded(slot) && next != nil {
 		b.logger.Trace("build block", "applying extrinsic", next)
 		ret, err := b.rt.ApplyExtrinsic(next)
+
+		t := b.transactionState.Pop()
 		if err != nil {
-			return nil, err
+			continue
 		}
```
Comment (noot) [inline lib/babe/build.go:191]: this should go above the previous line

File: lib/babe/build.go [hunk 5]
```diff
@@ -180,42 +177,38 @@ func (b *Service) buildBlockBabeHeader(slot Slot) (*types.BabeHeader, error) {
...
-func (b *Service) buildBlockExtrinsics(slot Slot) ([]*transaction.ValidTransaction, error) {
+func (b *Service) buildBlockExtrinsics(slot Slot) []*transaction.ValidTransaction {
 	next := b.nextReadyExtrinsic()
 	included := []*transaction.ValidTransaction{}
 
 	for !hasSlotEnded(slot) && next != nil {
 		b.logger.Trace("build block", "applying extrinsic", next)
 		ret, err := b.rt.ApplyExtrinsic(next)
+
+		t := b.transactionState.Pop()
```
Comment (noot) [inline lib/babe/build.go:188]: this should go above the previous line

File: lib/babe/build.go [hunk 6]
```diff
@@ -180,42 +177,38 @@ func (b *Service) buildBlockBabeHeader(slot Slot) (*types.BabeHeader, error) {
...
 		if err != nil {
-			return nil, err
+			continue
```
Comment (noot) [inline lib/babe/build.go:190]: add log here

File: lib/babe/build.go [hunk 7]
```diff
@@ -180,42 +177,40 @@ func (b *Service) buildBlockBabeHeader(slot Slot) (*types.BabeHeader, error) {
...
 	included := []*transaction.ValidTransaction{}
 
 	for !hasSlotEnded(slot) && next != nil {
 		b.logger.Trace("build block", "applying extrinsic", next)
+		t := b.transactionState.Pop()
 		ret, err := b.rt.ApplyExtrinsic(next)
+
 		if err != nil {
-			return nil, err
+			b.logger.Trace("apply extrinsic failed:", err, "extrinsic", next)
```
Comment (noot) [inline lib/babe/build.go:190]: can you add an `"error"` in from of the `err`? the logger expects an even number of arguments, after the initial info string

File: lib/babe/build.go [hunk 8]
```diff
@@ -180,42 +177,40 @@ func (b *Service) buildBlockBabeHeader(slot Slot) (*types.BabeHeader, error) {
...
-			}
 			// remove invalid extrinsic from queue
-			b.transactionState.Pop()
-
-			// re-add previously popped extrinsics back to queue
-			b.addToQueue(included)
-
-			return nil, errors.New("error applying extrinsic: " + errTxt)
+			if err == nil {
+				b.logger.Trace("failed to interpret extrinsic error", "extrinsic", next)
```
Comment (noot) [inline lib/babe/build.go:199]: could you add the `ret` error bytes here?

**Review by noot — APPROVED** [review]
looking good, please update logs!

**Review by arijitAD — APPROVED** [review]
Can you convert the `logger.Trace` to `logger.Error`? Otherwise looks good.

File: lib/babe/build.go (hunk 7 above)
Comment (RyRy79261) [reply lib/babe/build.go:190]: @noot 

Do you mean to replace the line as: 
```			b.logger.Trace("apply extrinsic failed:", "error", err, "extrinsic", next)
```

Comment (noot) [reply lib/babe/build.go:190]: yeah exactly, also change to `Warn` instead of `Trace`?

**Pull request merged**